Use SIMD in fast path for advance_for_char_range

bjwbell opened the pull request:
In advance_for_char_range add a fast SIMD code path for the the common
case where there are no detailed glyphs.

r? @mbrubeck

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7527)

<!-- Reviewable:end -->

bjwbell committed:
Add fast SIMD code path in advance_for_char_range

In advance_for_char_range add a SIMD code path for the the common
case of no detailed glyphs.

- gfx/Cargo.toml
  Add simd dependency https://github.com/huonw/simd

- servo/Cargo.lock
  Add simd dependency https://github.com/huonw/simd @
  d9ad79d86eab50a8f36d45fe17aa9e3a533389ee.

- SIMD isn't used on non-x86_65/aarch64 architectures.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

pcwalton commented on components/gfx/text/glyph.rs:
```diff
@@ -541,8 +547,41 @@ impl<'a> GlyphStore {
+        let len = rang.length().to_usize();
+        let num_simd_iterations = len / 4;
+        let leftover_entries = rang.end().to_usize() - (len - num_simd_iterations * 4);
+        let buf: &[u32] = unsafe { mem::transmute(self.entry_buffer.as_slice()) };
```
Can you factor this unsafe call out into a separate method and add a comment to the entry buffer field noting that if it is changed you may need to change the method too?

The safer way would be to use something like the nue crate here, but let's not pull in that dependency yet.

bjwbell replied:
Sorry for the delay. I factored it out today.

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit efd3e08 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton efd3e08ed43dc200bc339797c3d0f2016c565bf4 -->

bors-servo commented:
:hourglass: Testing commit efd3e08 with merge 9daaa85...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/2184)

bjwbell commented:
Looks like the SIMD crate fails to build on android due to an unused import (arm::neon::common). I'll look at it later today (have to run out for couple hours).

huonw commented:
The problem here is NEON isn't guaranteed to exist on all ARM devices, meaning `rustc` (and hence `simd`) cannot assume it exists for the default `arm-...` targets. As you've found out, the crate handles these cases particularly badly. (It also affects 32-bit x86 (the `i686-...`) targets at the moment.)

One way to fix is to use a wrapper script around rustc (e.g. like `etc/rustc-with-gold`) that passes `-C target-feature=+neon` on the arm targets, if you're OK with only running on devices with NEON support.

bjwbell commented:
For now, I'll disable SIMD in servo on ARM. And put in a note in the code that for ARM it'll always use the slow path.  

I'd rather do the wrapper script as a separate PR, since I've never modified the servo build process.

bjwbell commented:
Having https://github.com/rust-lang/cargo/issues/1137 (Expose target-cpu/target-feature) would be really nice for this.

bjwbell commented:
@pcwalton can you have bors rebuild & test? I can't do it locally (only have an x86_64).

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 3c04103 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 3c04103a1397f34b7624ce1a4f0fcbb5d867b076 -->

bors-servo commented:
:hourglass: Testing commit 3c04103 with merge c06636c...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/2216)

bjwbell commented:
Third times the charm.

I changed the simd dependency in gfx/Cargo.toml to be target triple specific and added target triples
for:
- x86_64-unknown-linux-gnu
- x86_64-apple-darwin
- aarch64-unknown-linux-gnu
- x86_64-pc-windows-gnu
- x86_64-pc-windows-msvc

No idea, if the android builder's target triple is in the above list. 

@pcwalton you can retry bors.

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 069f9d7 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 069f9d74b46c6afdb052cf52fa9fbc97b65d6505 -->

bors-servo commented:
:hourglass: Testing commit 069f9d7 with merge bd9f579...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/2218)

metajack commented:
The android triple is arm-linux-androideabi

bjwbell commented:
@metajack thanks, I realized I only need it for aarch64 on android since the simd crate isn't used for any of the other arm android platforms.

bjwbell commented:
@metajack Is the android aarch64 triple aarch64-linux-androideabi? If it is I can add it.

metajack commented:
@bjwbell I don't actually know. @larsbergstrom ?

metajack commented:
@bors-servo r+

---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 1 of 1 files at r5.
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7527)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 0c99b7e has been approved by `metajack`

<!-- @bors-servo r=metajack 0c99b7e6a9cde248f66637c4d7c8ce8b51ce5c14 -->

bors-servo commented:
:hourglass: Testing commit 0c99b7e with merge c838259...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/3059), [gonk](http://build.servo.org/builders/gonk/builds/2219), [linux-dev](http://build.servo.org/builders/linux-dev/builds/145), [linux-rel](http://build.servo.org/builders/linux-rel/builds/143), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/145), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/144), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/153)

larsbergstrom commented:
@metajack @bjwbell Yes, that appears correct. For SIMD/NEON, I think we also need to figure out if we should be targeting `armv7-linux-androideabi`, since NEON is available in armv7, too, IIRC.
